analytics/pubstack/pubstack_module.go

```diff
@@ -0,0 +1,276 @@
+	}
+	if cfg.Features[VIDEO] {
+		newChanMap[VIDEO] = eventchannel.NewChannel(cfg.Endpoint, VIDEO, p.buffsCfg.size, p.buffsCfg.count, p.buffsCfg.timeout)
+	}
+	if cfg.Features[SETUID] {
+		newChanMap[SETUID] = eventchannel.NewChannel(cfg.Endpoint, SETUID, p.buffsCfg.size, p.buffsCfg.count, p.buffsCfg.timeout)
+	}
+
+	p.chans = newChanMap
+	p.cfg = cfg
+}
+
+func (p *PubstackModule) LogAuctionObject(ao *analytics.AuctionObject) {
+	// check if we have to send auctions events
+	ch, ok := p.chans[AUCTION]
+	if !ok {
+		return
+	}
+
+	// serialize event
+	payload, err := helpers.JsonifyAuctionObject(ao, p.scope)
+	if err != nil {
+		glog.Warning("Cannot serialize auction")
+		return
+	}
+
+	ch.Add(payload)
+}
+
+func (p *PubstackModule) LogVideoObject(vo *analytics.VideoObject) {
+	// check if we have to send auctions events
+	ch, ok := p.chans[VIDEO]
+	if !ok {
+		return
+	}
+
+	// serialize event
+	payload, err := helpers.JsonifyVideoObject(vo, p.scope)
+	if err != nil {
+		glog.Warning("Cannot serialize video")
+		return
+	}
+
+	ch.Add(payload)
+}
+
+func (p *PubstackModule) LogSetUIDObject(so *analytics.SetUIDObject) {
+	// check if we have to send auctions events
+	ch, ok := p.chans[SETUID]
+	if !ok {
+		return
+	}
+
+	// serialize event
+	payload, err := helpers.JsonifySetUIDObject(so, p.scope)
+	if err != nil {
+		glog.Warning("Cannot serialize video")
+		return
+	}
+
+	ch.Add(payload)
+}
+
+func (p *PubstackModule) LogCookieSyncObject(cso *analytics.CookieSyncObject) {
+	// check if we have to send auctions events
+	ch, ok := p.chans[COOKIESYNC]
+	if !ok {
+		return
+	}
+
+	// serialize event
+	payload, err := helpers.JsonifyCookieSync(cso, p.scope)
+	if err != nil {
+		glog.Warning("Cannot serialize video")
+		return
+	}
+
+	ch.Add(payload)
+}
+
+func (p *PubstackModule) LogAmpObject(ao *analytics.AmpObject) {
+	// check if we have to send auctions events
+	ch, ok := p.chans[AMP]
+	if !ok {
+		return
+	}
+
+	// serialize event
+	payload, err := helpers.JsonifyAmpObject(ao, p.scope)
+	if err != nil {
+		glog.Warning("Cannot serialize video")
+		return
+	}
+
+	ch.Add(payload)
+}
+
+func (p *PubstackModule) refreshConfiguration(waitPeriod time.Duration, end chan os.Signal) {
+	tick := time.NewTicker(waitPeriod)
+
+	for {
+		select {
+		case <-tick.C:
+			config, err := getConfiguration(p.cfg.ScopeId, p.cfg.Endpoint)
+			if err != nil {
+				glog.Error("fail to update configuration")
+				continue
+			}
+			p.applyConfiguration(config)
+		case <-end:
+			return
+		}
+	}
+}
+
+func getConfiguration(scope string, intake string) (*Configuration, error) {
+	u, err := url.Parse(intake)
+	if err != nil {
+		return nil, err
+	}
+
+	u.Path = path.Join(u.Path, "bootstrap")
+	q, _ := url.ParseQuery(u.RawQuery)
+
+	q.Add("scopeId", scope)
+	u.RawQuery = q.Encode()
+
+	res, err := http.DefaultClient.Get(u.String())
+	if err != nil {
+		return nil, err
+	}
+
+	defer res.Body.Close()
+	data, err := ioutil.ReadAll(res.Body)
```

Comment:
Please use `json.Decoder` for unmarshaling from a stream. See: https://stackoverflow.com/questions/21197239/decoding-json-using-json-unmarshal-vs-json-newdecoder-decode?rq=1